File: crates/netssh-python/textfsm/parse_output.py
#!/usr/bin/env python3
"""
TextFSM command output parsing module.

This module provides functions to parse command outputs using TextFSM templates.
It handles template location, selection, and output parsing.
"""

import os
import re
import csv
import logging
import json
from typing import List, Dict, Any, Optional, Union

# Setup logging
logger = logging.getLogger(__name__)

# Path to templates
TEMPLATE_DIR = os.path.join(os.path.dirname(os.path.dirname(__file__)), "textfsm/templates")

# Import TextFSM
from .parse import TextFSM

class NetworkOutputParser:
    """Class for parsing network device command outputs using TextFSM templates."""
    
    def __init__(self, template_dir=None):
        """
        Initialize the parser with a template directory.
    
    Args:
            template_dir: Optional path to template directory. If not provided,
                          uses the default template directory.
        """
        self.template_dir = template_dir or TEMPLATE_DIR
        self._index_loaded = False
        self._platform_templates = {}
        
    def _load_template_index(self):
        """Load the template index file and build a dictionary of templates by platform."""
        if self._index_loaded:
            return self._platform_templates
            
        index_file = os.path.join(self.template_dir, "index")
    if not os.path.isfile(index_file):
        logger.error(f"Index file not found at {index_file}")
            self._index_loaded = True
            return {}
            
        try:
            # Read the index file
            with open(index_file, 'r') as f:
                # Skip comment lines at the beginning
                header_found = False
                header_line = None
                for line in f:
                    line = line.strip()
                    if not line or line.startswith('#'):
                        continue
                    
                    # First non-comment line is the header
                    if not header_found:
                        header_line = line
                        header_found = True
                        break
                
                if not header_found:
                    logger.error("No header found in index file")
                    self._index_loaded = True
                    return {}
                
                # Parse the header to get column positions
                if not header_line:
                    logger.error("No valid header line found in index file")
                    self._index_loaded = True
                    return {}
                
                headers = [h.strip() for h in header_line.split(',')]
                
                # Find the column indices for template, platform, and command
                template_col = -1
                platform_col = -1
                command_col = -1
                
                for i, header in enumerate(headers):
                    if header.lower() == 'template':
                        template_col = i
                    elif header.lower() in ['platform', 'vendor']:
                        platform_col = i
                    elif header.lower() == 'command':
                        command_col = i
                
                if template_col == -1 or platform_col == -1 or command_col == -1:
                    logger.error(f"Missing required columns in index file. Headers: {headers}")
                    self._index_loaded = True
                    return {}
                
                # Reset file pointer to beginning
                f.seek(0)
                
                # Skip header and comments
                header_passed = False
                for line in f:
                    line = line.strip()
                    if not line or line.startswith('#'):
                        continue
                    
                    if not header_passed:
                        header_passed = True
                        continue
                    
                    # Parse the line using CSV parser to handle quoted fields
                    try:
                        row = next(csv.reader([line]))
                        
                        if len(row) <= max(template_col, platform_col, command_col):
                        continue
                        
                        template_file = row[template_col].strip()
                        platform = row[platform_col].strip().lower()
                        command_pattern = row[command_col].strip()
                        
                        # Remove the regex patterns for command completion
                        # Convert sh[[ow]] ver[[sion]] to show version
                        command = re.sub(r'\[\[([^]]+)\]\]', r'\1', command_pattern).lower()
                        
                        # Add to platform templates
                        if platform not in self._platform_templates:
                            self._platform_templates[platform] = []
                        
                        self._platform_templates[platform].append({
                            'command': command,
                            'template': template_file
                        })
                    except Exception as e:
                        logger.debug(f"Error parsing line '{line}': {str(e)}")
                                    continue
                                    
        except Exception as e:
            logger.error(f"Error loading template index: {str(e)}")
            
        self._index_loaded = True
        logger.debug(f"Loaded templates for {len(self._platform_templates)} platforms")
        return self._platform_templates
    
    def find_template(self, platform, command):
        """
        Find the appropriate template for a given platform and command.
        
        Args:
            platform: Device platform (e.g., cisco_ios)
            command: Command string (e.g., show version)
            
        Returns:
            Path to template file or None if not found
        """
        templates = self._load_template_index()
        platform = platform.lower()
        command = command.lower()
        
        if platform not in templates:
            logger.warning(f"No templates found for platform '{platform}'")
            return None
            
        logger.debug(f"Looking for command '{command}' in platform '{platform}' with {len(templates[platform])} templates")
        
        # First try exact match
        for entry in templates[platform]:
            if command == entry['command']:
                template_path = os.path.join(self.template_dir, entry['template'])
                if os.path.isfile(template_path):
                    logger.debug(f"Found exact match template: {template_path}")
                                return template_path
                                
        # Then try substring match
        for entry in templates[platform]:
            if command in entry['command'] or entry['command'] in command:
                template_path = os.path.join(self.template_dir, entry['template'])
                if os.path.isfile(template_path):
                    logger.debug(f"Found substring match template: {template_path}")
                                return template_path
                                
        logger.warning(f"No template found for '{platform}' command '{command}'")
        return None
    
    def parse_output(self, platform, command, data):
    """
    Parse command output using TextFSM.
    
    Args:
            platform: Device platform (e.g., cisco_ios)
            command: Command string (e.g., show version)
            data: Command output as string
        
    Returns:
            List of dictionaries containing parsed data, or None if parsing fails
    """
        if not data:
        logger.warning("Empty output provided for parsing")
        return None
            
        template_path = self.find_template(platform, command)
        if not template_path:
            logger.warning(f"No template found for {platform}, {command}")
            return None
    
    try:
        logger.debug(f"Parsing output using template: {template_path}")
        with open(template_path, 'r') as template_file:
                fsm = TextFSM(template_file)
                parsed_data = fsm.ParseTextToDicts(data)
                print("+++++++++++++++++++++++++++++++++++++++++++++++++++++")
                print(parsed_data)
                print("+++++++++++++++++++++++++++++++++++++++++++++++++++++")
                # Convert to list of dictionaries
                result = []
                for record in parsed_data:
                    record_dict = {}
                    for i, header in enumerate(fsm.header):
                        record_dict[header] = record[i]
                    result.append(record_dict)
                
                return result
    except Exception as e:
        logger.error(f"Error parsing output: {str(e)}")
        return None

    def parse_to_json(self, platform, command, data):
        """
        Parse command output and return as JSON string.
        
        Args:
            platform: Device platform (e.g., cisco_ios)
            command: Command string (e.g., show version)
            data: Command output as string
            
        Returns:
            JSON string of parsed data, or None if parsing fails
        """
        result = self.parse_output(platform, command, data)
        if result:
            return json.dumps(result, indent=2)
        return None

# For backward compatibility
parser = NetworkOutputParser()

def parse_output(platform, command, data):
    """
    Parse command output using TextFSM (function interface).
    
    Args:
        platform: Device platform (e.g., cisco_ios)
        command: Command string (e.g., show version)
        data: Command output as string
        
    Returns:
        List of dictionaries containing parsed data, or None if parsing fails
    """
    return parser.parse_output(platform, command, data)

def parse_output_to_json(platform, command, data):
    """
    Parse command output and return as JSON string (function interface).
    
    Args:
        platform: Device platform (e.g., cisco_ios)
        command: Command string (e.g., show version)
        data: Command output as string
        
    Returns:
        JSON string of parsed data, or None if parsing fails
    """
    return parser.parse_to_json(platform, command, data)

# Make these functions available when imported
__all__ = ['parse_output', 'parse_output_to_json', 'NetworkOutputParser'] 

File: crates/netssh-textfsm/src/parse_output.rs
use crate::textfsm::TextFSM;
use indexmap::IndexMap;
use regex::Regex;
use serde_json;
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::PathBuf;
use std::sync::{Arc, Mutex};

/// Custom error type for parse output operations
pub type ParseOutputResult<T> = std::result::Result<T, Box<dyn std::error::Error>>;

/// Compiled template entry with pre-compiled regex and validated path
#[derive(Debug, Clone)]
struct CompiledTemplateEntry {
    /// Pre-compiled command regex pattern
    command_regex: Option<Regex>,
    /// Original command pattern for fallback
    command_pattern: String,
    /// Full path to template file (pre-validated)
    template_path: PathBuf,
}

/// Global compiled index cache
#[derive(Debug, Clone)]
struct CompiledIndex {
    /// Platform to compiled templates mapping
    platform_templates: HashMap<String, Vec<CompiledTemplateEntry>>,
}

lazy_static::lazy_static! {
    /// Pre-compiled regex for command pattern normalization
    static ref COMPLETION_PATTERN: Regex = Regex::new(r"\[\[([^\]]+)\]\]").unwrap();

    /// Global cache for compiled indexes
    static ref GLOBAL_INDEX_CACHE: Arc<Mutex<HashMap<String, CompiledIndex>>> =
        Arc::new(Mutex::new(HashMap::new()));
}

/// Get global index cache
fn get_global_cache() -> Arc<Mutex<HashMap<String, CompiledIndex>>> {
    GLOBAL_INDEX_CACHE.clone()
}

/// Convert completion patterns to regex patterns
/// Converts "sh[[ow]]" to "sh(o(w)?)?" and "ver[[sion]]" to "ver(s(i(o(n)?)?)?)?)"
fn convert_completion_to_regex(pattern: &str) -> String {
    COMPLETION_PATTERN.replace_all(pattern, |caps: &regex::Captures| {
        let word = &caps[1];
        let mut result = String::new();
        for ch in word.chars() {
            result.push('(');
            result.push(ch);
        }
        result.push_str(&")?".repeat(word.len()));
        result
    }).to_string()
}

/// Normalize command pattern by removing completion brackets and converting to lowercase
fn normalize_command_pattern(pattern: &str) -> String {
    COMPLETION_PATTERN.replace_all(pattern, "$1").to_lowercase()
}

/// Network output parser for TextFSM templates
///
/// This struct provides functionality to parse network device command outputs
/// using TextFSM templates. It handles template location, selection, and output parsing.
#[derive(Debug)]
pub struct NetworkOutputParser {
    /// Path to the templates directory
    template_dir: PathBuf,
}

impl NetworkOutputParser {
    /// Create a new NetworkOutputParser
    ///
    /// # Arguments
    /// * `template_dir` - Optional path to template directory. If None, uses default.
    pub fn new(template_dir: Option<PathBuf>) -> Self {
        let default_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("templates");

        Self {
            template_dir: template_dir.unwrap_or(default_dir),
        }
    }

    /// Get or load the compiled index for this template directory
    fn get_compiled_index(&self) -> ParseOutputResult<CompiledIndex> {
        let cache = get_global_cache();
        let cache_key = self.template_dir.to_string_lossy().to_string();

        // Check if already cached
        {
            let cache_guard = cache.lock().unwrap();
            if let Some(compiled_index) = cache_guard.get(&cache_key) {
                return Ok(compiled_index.clone());
            }
        }

        // Load and compile index
        let compiled_index = self.load_and_compile_index()?;

        // Cache the compiled index
        {
            let mut cache_guard = cache.lock().unwrap();
            cache_guard.insert(cache_key, compiled_index.clone());
        }

        Ok(compiled_index)
    }

    /// Load and compile the template index file
    fn load_and_compile_index(&self) -> ParseOutputResult<CompiledIndex> {
        let index_file = self.template_dir.join("index");
        if !index_file.exists() {
            eprintln!("Index file not found at {}", index_file.display());
            return Ok(CompiledIndex {
                platform_templates: HashMap::new(),
            });
        }

        let file = File::open(&index_file)?;
        let reader = BufReader::new(file);
        let mut lines = reader.lines();

        // Find the header line (first non-comment line)
        let mut header_line = None;
        for line in lines.by_ref() {
            let line = line?;
            let trimmed = line.trim();

            if trimmed.is_empty() || trimmed.starts_with('#') {
                continue;
            }

            header_line = Some(trimmed.to_string());
            break;
        }

        let header = header_line.ok_or("No header found in index file")?;

        // Parse header to get column positions
        let headers: Vec<&str> = header.split(',').map(|h| h.trim()).collect();

        let template_col = headers.iter().position(|&h| h.to_lowercase() == "template")
            .ok_or("Missing 'template' column in index file")?;
        let platform_col = headers.iter().position(|&h| {
            let lower = h.to_lowercase();
            lower == "platform" || lower == "vendor"
        }).ok_or("Missing 'platform' or 'vendor' column in index file")?;
        let command_col = headers.iter().position(|&h| h.to_lowercase() == "command")
            .ok_or("Missing 'command' column in index file")?;

        let mut platform_templates: HashMap<String, Vec<CompiledTemplateEntry>> = HashMap::new();

        // Parse the remaining lines
        for line in lines {
            let line = line?;
            let trimmed = line.trim();

            if trimmed.is_empty() || trimmed.starts_with('#') {
                continue;
            }

            // Parse CSV line (simple split for now, could be enhanced for quoted fields)
            let fields: Vec<&str> = trimmed.split(',').map(|f| f.trim()).collect();

            if fields.len() <= template_col.max(platform_col).max(command_col) {
                continue;
            }

            let template_files = fields[template_col];
            let platform = fields[platform_col].to_lowercase();
            let command_pattern = fields[command_col];

            // Handle multiple template files separated by colons - use only the first one
            let template_file = template_files.split(':').next().unwrap_or(template_files).to_string();

            // Build full template path and validate existence
            let template_path = self.template_dir.join(&template_file);
            if !template_path.exists() {
                // Skip silently - some templates may not exist in this distribution
                continue;
            }

            // Convert completion patterns to regex and compile
            let regex_pattern = convert_completion_to_regex(command_pattern);
            let command_regex = match Regex::new(&regex_pattern) {
                Ok(regex) => Some(regex),
                Err(_) => {
                    // Skip patterns that can't be compiled (e.g., lookahead/lookbehind)
                    None
                }
            };

            let entry = CompiledTemplateEntry {
                command_regex,
                command_pattern: command_pattern.to_string(),
                template_path,
            };

            platform_templates
                .entry(platform)
                .or_insert_with(Vec::new)
                .push(entry);
        }


        Ok(CompiledIndex {
            platform_templates,
        })
    }

    /// Find the appropriate template for a given platform and command
    ///
    /// # Arguments
    /// * `platform` - Device platform (e.g., "cisco_ios")
    /// * `command` - Command string (e.g., "show version")
    ///
    /// # Returns
    /// Path to template file or None if not found
    pub fn find_template(&self, platform: &str, command: &str) -> ParseOutputResult<Option<PathBuf>> {
        let compiled_index = self.get_compiled_index()?;
        let platform_lower = platform.to_lowercase();
        let command_lower = command.to_lowercase();

        let platform_entries = match compiled_index.platform_templates.get(&platform_lower) {
            Some(entries) => entries,
            None => {
                return Ok(None);
            }
        };

        // First try pre-compiled regex match (fast!)
        for entry in platform_entries {
            if let Some(ref regex) = entry.command_regex {
                if regex.is_match(&command_lower) {
                    return Ok(Some(entry.template_path.clone()));
                }
            }
        }

        // Fallback to substring match for patterns that couldn't be compiled as regex
        for entry in platform_entries {
            let normalized_pattern = normalize_command_pattern(&entry.command_pattern);
            if command_lower.contains(&normalized_pattern) || normalized_pattern.contains(&command_lower) {
                return Ok(Some(entry.template_path.clone()));
            }
        }

        Ok(None)
    }

    /// Parse command output using TextFSM
    ///
    /// # Arguments
    /// * `platform` - Device platform (e.g., "cisco_ios")
    /// * `command` - Command string (e.g., "show version")
    /// * `data` - Command output as string
    ///
    /// # Returns
    /// Vector of dictionaries containing parsed data, or None if parsing fails
    pub fn parse_output(
        &self,
        platform: &str,
        command: &str,
        data: &str,
    ) -> ParseOutputResult<Option<Vec<IndexMap<String, serde_json::Value>>>> {
        if data.is_empty() {
            return Ok(None);
        }

        let template_path = match self.find_template(platform, command)? {
            Some(path) => path,
            None => {
                return Ok(None);
            }
        };

        let template_file = File::open(&template_path)?;

        let mut fsm = TextFSM::new(template_file)?;
        let parsed_data = fsm.parse_text_to_dicts(data)?;

        Ok(Some(parsed_data))
    }

    /// Parse command output and return as JSON string
    ///
    /// # Arguments
    /// * `platform` - Device platform (e.g., "cisco_ios")
    /// * `command` - Command string (e.g., "show version")
    /// * `data` - Command output as string
    ///
    /// # Returns
    /// JSON string of parsed data, or None if parsing fails
    pub fn parse_to_json(
        &self,
        platform: &str,
        command: &str,
        data: &str,
    ) -> ParseOutputResult<Option<String>> {
        match self.parse_output(platform, command, data)? {
            Some(result) => {
                let json = serde_json::to_string_pretty(&result)?;
                Ok(Some(json))
            }
            None => Ok(None),
        }
    }
}

// Global parser instance for backward compatibility
lazy_static::lazy_static! {
    static ref GLOBAL_PARSER: std::sync::Mutex<NetworkOutputParser> = 
        std::sync::Mutex::new(NetworkOutputParser::new(None));
}

/// Parse command output using TextFSM (function interface)
///
/// # Arguments
/// * `platform` - Device platform (e.g., "cisco_ios")
/// * `command` - Command string (e.g., "show version")
/// * `data` - Command output as string
///
/// # Returns
/// Vector of dictionaries containing parsed data, or None if parsing fails
pub fn parse_output(
    platform: &str,
    command: &str,
    data: &str,
) -> ParseOutputResult<Option<Vec<IndexMap<String, serde_json::Value>>>> {
    let parser = GLOBAL_PARSER.lock().unwrap();
    parser.parse_output(platform, command, data)
}

/// Parse command output and return as JSON string (function interface)
///
/// # Arguments
/// * `platform` - Device platform (e.g., "cisco_ios")
/// * `command` - Command string (e.g., "show version")
/// * `data` - Command output as string
///
/// # Returns
/// JSON string of parsed data, or None if parsing fails
pub fn parse_output_to_json(
    platform: &str,
    command: &str,
    data: &str,
) -> ParseOutputResult<Option<String>> {
    let parser = GLOBAL_PARSER.lock().unwrap();
    parser.parse_to_json(platform, command, data)
}
